lib/stdlib/src/timer.erl

```diff
@@ -236,106 +278,128 @@ hours(Hours) ->
-    BRef = {Started + 1000*Time, make_ref()},
-    Timer = {BRef, timeout, Op},
-    ets:insert(?TIMER_TAB, Timer),
-    Timeout = timer_timeout(system_time()),
-    {reply, {ok, BRef}, [], Timeout};
-handle_call({apply_interval, {Time, To, MFA}, Started}, _From, _Ts) 
-  when is_integer(Time), Time >= 0 ->
-    %% To must be a pid or a registered name
-    case get_pid(To) of
-	Pid when is_pid(Pid) ->
-	    catch link(Pid),
-	    SysTime = system_time(),
-	    Ref = make_ref(),
-	    BRef1 = {interval, Ref},
-	    Interval = Time*1000,
-	    BRef2 = {Started + Interval, Ref},
-	    Timer = {BRef2, {repeat, Interval, Pid}, MFA},
-	    ets:insert(?INTERVAL_TAB, {BRef1,BRef2,Pid}),
-	    ets:insert(?TIMER_TAB, Timer),
-	    Timeout = timer_timeout(SysTime),
-	    {reply, {ok, BRef1}, [], Timeout};
-	_ ->
-	    {reply, {error, badarg}, [], next_timeout()}
-    end;
-handle_call({cancel, BRef = {_Time, Ref}, _}, _From, Ts) 
-  when is_reference(Ref) ->
-    delete_ref(BRef),
-    {reply, {ok, cancel}, Ts, next_timeout()};
-handle_call({cancel, _BRef, _}, _From, Ts) ->
-    {reply, {error, badarg}, Ts, next_timeout()};
-handle_call({apply_after, _, _}, _From, Ts) ->
-    {reply, {error, badarg}, Ts, next_timeout()};
-handle_call({apply_interval, _, _}, _From, Ts) ->
-    {reply, {error, badarg}, Ts, next_timeout()};
-handle_call(_Else, _From, Ts) ->		  % Catch anything else
-    {noreply, Ts, next_timeout()}.
-
--spec handle_info(term(), timers()) -> {'noreply', timers(), timeout()}.
-handle_info(timeout, Ts) ->                       % Handle timeouts 
-    Timeout = timer_timeout(system_time()),
-    {noreply, Ts, Timeout};
-handle_info({'EXIT',  Pid, _Reason}, Ts) ->       % Oops, someone died
-    pid_delete(Pid),
-    {noreply, Ts, next_timeout()};
-handle_info(_OtherMsg, Ts) ->                     % Other Msg's
-    {noreply, Ts, next_timeout()}.
-
--spec handle_cast(term(), timers()) -> {'noreply', timers(), timeout()}.
-handle_cast(_Req, Ts) ->                          % Not predicted but handled
-    {noreply, Ts, next_timeout()}.
+%% Call handling.
+-spec handle_call(term(), term(), #state{}) ->
+        {'reply', term(), #state{}, timeout()} | {'noreply', #state{}, timeout()}.
+%% Start an one-shot timer.
+handle_call({apply_after, {Time, MFA}, Started}, _From, State) ->
+    StartTime = Started + 1000 * Time,
+    TRef = {StartTime, make_ref()},
+    Timer = {TRef, timeout, MFA},
+    ets:insert(State#state.timer_tab, Timer),
+    {reply, {ok, TRef}, State, next_timeout(State)};
+%% Start an interval timer.
+handle_call({apply_interval, {Time, Pid, MFA}, Started}, _From, State) ->
+    Interval = Time * 1000,
+    StartTime = Started + Interval,
+    Ref = monitor(process, Pid),
+    TRef = {StartTime, Ref},
+    Timer = {TRef, {repeat, Interval}, MFA},
+    ets:insert(State#state.interval_tab, {Ref, TRef}),
+    ets:insert(State#state.timer_tab, Timer),
+    {reply, {ok, {interval, Ref}}, State, next_timeout(State)};
```

Comment:
I wonder if this implementation is trying to be too smart because it was implemented before timer wheels were added to the VM. Maybe the simplest and faster approach now is to use send_after instead of getting the `next_timeout`. This would have the benefit that the management of the timer for operations like `apply_after` could be done completely in the client (the server would effectively only receive the MFA on timeout).

Reply:
Hm, not a bad idea... I'll have to think this through a bit more, but in theory, this could eliminate the need for the `ordered_set` timer table altogether. Interval timers would be just timers that reschedule the send, too. It will probably still all have to go through the timer server, so we know what timers we _actually_ have and for things like `cancel` etc, but... yeah :)

@Maria-12648430 any objections, in terms of keeping outward behavior specifically?

Reply:
> @Maria-12648430 any objections, in terms of keeping outward behavior specifically?

Not off the top of my head, no. But I will check in more depth, after lunch :)

Reply:
No objections, at least I don't see any problems ^^;

Reply:
By using the `{abs, true}` option of `erlang:send_after/4` or `erlang:start_timer/4` you can prevent that interval timers slowly drifts as well.

Reply:
Had a quick look at the code and slow drift of intervals is perhaps not an issue in current implementation. Anyway, the `{abs, true}` option may help with intervals if rewritten using `send_after()`/`start_timer()`.